compiler/rustc_monomorphize/src/collector/abi_check.rs

```diff
@@ -0,0 +1,201 @@
+        match terminator.kind {
+            mir::TerminatorKind::Call { ref func, ref fn_span, .. }
+            | mir::TerminatorKind::TailCall { ref func, ref fn_span, .. } => {
+                let callee_ty = func.ty(self.body, self.tcx);
+                let callee_ty = self.instance.instantiate_mir_and_normalize_erasing_regions(
+                    self.tcx,
+                    ty::ParamEnv::reveal_all(),
+                    ty::EarlyBinder::bind(callee_ty),
+                );
+                check_call_site_abi(self.tcx, callee_ty, *fn_span, self.body.source.instance);
+            }
+            _ => {}
+        }
+    }
+}
+
+fn check_callees_abi<'tcx>(tcx: TyCtxt<'tcx>, instance: Instance<'tcx>) {
+    let body = tcx.instance_mir(instance.def);
+    let mut visitor = MirCallesAbiCheck { tcx, body, instance };
+    for (bb, data) in traversal::mono_reachable(body, tcx, instance) {
+        visitor.visit_basic_block_data(bb, data)
+    }
+}
+
+fn should_check_instance_abi<'tcx>(tcx: TyCtxt<'tcx>, instance: Instance<'tcx>) -> bool {
+    // We do not need to check the instance for feature-dependent ABI if we can determine that the
+    // function has "Rust" or "rust-call" ABI, which is known not to be feature-dependent.
+    // Note that the check is still correct on Rust ABI functions, but somewhat expensive. Hence,
+    // checking for "Rust" ABI is just an optimization.
+    // We also avoid to try to determine the type of the instance, as doing so involves running a
+    // query that does not usually run for unchanged functions in incremental builds.
+    match instance.def {
+        // We only need to check for user-defined functions - if all user-defined functions are
+        // fine, so are the `instance`s derived by the compiler.
+        InstanceKind::Item(def) => {
+            // fn_sig ICEs on defs that are not functions.
+            if matches!(tcx.def_kind(def), DefKind::Fn | DefKind::AssocFn) {
+                !tcx.fn_sig(def).skip_binder().abi().is_rust()
+            } else if matches!(tcx.def_kind(def), DefKind::Ctor(..) | DefKind::Closure) {
+                // Struct constructors and closures do not give control of their ABI to the user.
+                false
+            } else {
+                true
+            }
+        }
+        InstanceKind::ReifyShim(..)
+        | InstanceKind::FnPtrShim(..)
+        | InstanceKind::Virtual(..)
+        | InstanceKind::VTableShim(..)
+        | InstanceKind::Intrinsic(..)
+        | InstanceKind::DropGlue(..)
+        | InstanceKind::CloneShim(..)
+        | InstanceKind::FnPtrAddrShim(..)
+        | InstanceKind::AsyncDropGlueCtorShim(..)
+        | InstanceKind::ThreadLocalShim(..)
+        | InstanceKind::ClosureOnceShim { .. }
+        | InstanceKind::ConstructCoroutineInClosureShim { .. } => false,
+    }
+}
+
+fn check_feature_dependent_abi<'tcx>(tcx: TyCtxt<'tcx>, instance: Instance<'tcx>) {
+    if should_check_instance_abi(tcx, instance) {
```

Comment:
I doubt this optimization still has a measurable effect.

Reply:
Agreed!
However, it has the effect of not messing with the query order in the collector in most cases...
I assume we'd prefer to skip the "optimization" though, and be more confident in the correctness of the check...

Reply:
Yeah I'd prefer that.